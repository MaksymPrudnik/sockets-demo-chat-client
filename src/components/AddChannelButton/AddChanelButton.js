import React, { useState } from "react";
import { socket } from "../../service/socket";
import styled from "styled-components";

const AddChannelDiv = styled.div`
  display: flex;
  margin: 0.5rem 1rem 0;
`;
const ChannelName = styled.input`
  background-color: ${({ theme }) => theme.body};
  color: ${({ theme }) => theme.color};
  padding: 0.3rem 0.5rem;
  border: 1.5px solid ${({ theme }) => theme.color};
  border-top-left-radius: 0.7rem;
  border-bottom-left-radius: 0.7rem;
`;
const AddButton = styled.button`
  background-color: ${({ theme }) => theme.body};
  color: ${({ theme }) => theme.color};
  font-weight: 500;
  text-transform: uppercase;
  border: 1.5px solid ${({ theme }) => theme.color};
  border-top-right-radius: 0.7rem;
  border-bottom-right-radius: 0.7rem;
`;

const AddChannelButton = () => {
  const [channelName, setChannelName] = useState("");
  const handleNameChange = (e) => {
    setChannelName(e.target.value);
  };
  const handleAddChannel = () => {
    if (!channelName) return;
    socket.emit("new channel", channelName.replace(" ", "-"));
    setChannelName("");
  };
  const handleCheckChars = (e) => {
    const allowedChars = /[0-9a-zA-Z\ ]/;
    if (!allowedChars.test(e.data) || (!channelName.length && e.data === " "))
      e.preventDefault();
  };
  return (
    <AddChannelDiv>
      <ChannelName
        type="text"
        placeholder="Channel name"
        value={channelName}
        onChange={handleNameChange}
        onBeforeInput={handleCheckChars}
      />
      <AddButton onClick={handleAddChannel}>add</AddButton>
    </AddChannelDiv>
  );
};

export default AddChannelButton;
