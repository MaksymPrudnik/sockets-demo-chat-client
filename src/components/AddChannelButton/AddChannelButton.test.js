import React from 'react';
import { shallow } from 'enzyme';
import AddChannelButton from './AddChanelButton';

describe('<AddChannelButton />', () => {
    const wrapper = shallow(<AddChannelButton />);
    it('should match the snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot();
    })

    it('should have a text field', () => {
        expect(wrapper.find('input[type="text"]').length).toEqual(1);
    })

    it('[text input] should prevent invalid chars input', () => {
        const beforeInputParams = {
            data: 'a',
            preventDefault: jest.fn(() => {})
        }
        wrapper.find('input[type="text"]').simulate('beforeinput', beforeInputParams)
        expect(beforeInputParams.preventDefault).toHaveBeenCalledTimes(0)
        wrapper.find('input[type="text"]').simulate('beforeinput', {...beforeInputParams, data: '-'})
        expect(beforeInputParams.preventDefault).toHaveBeenCalledTimes(1)
        wrapper.find('input[type="text"]').simulate('beforeinput', {...beforeInputParams, data: ' '})
        expect(beforeInputParams.preventDefault).toHaveBeenCalledTimes(2)
        wrapper.find('input[type="text"]').simulate('change', { target: { value: 'm' } } )
        wrapper.find('input[type="text"]').simulate('beforeinput', {...beforeInputParams, data: ' '})
        expect(beforeInputParams.preventDefault).toHaveBeenCalledTimes(2)
    })

    it('[text input] should update onChange', () => {
        wrapper.find('input[type="text"]').simulate('change', { target: { value: 'm' } } )
        expect(wrapper.find('input[type="text"]').prop('value')).toEqual('m');
        wrapper.find('input[type="text"]').simulate('change', { target: { value: 'new channel' } } )
        expect(wrapper.find('input[type="text"]').prop('value')).toEqual('new channel');
    })

    it('[button] should clear text input', () => {
        wrapper.find('input[type="text"]').simulate('change', { target: { value: 'new channel' } } )
        wrapper.find('button').simulate('click')
        expect(wrapper.find('input[type="text"]').prop('value')).toEqual('');
    })
})