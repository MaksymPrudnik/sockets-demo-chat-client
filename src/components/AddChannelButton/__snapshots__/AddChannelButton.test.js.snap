// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddChannelButton /> should match the snapshot 1`] = `"<div class=\\"add-channel-button\\"><input type=\\"text\\" placeholder=\\"Chanel name\\" value=\\"\\"/><button>+</button></div>"`;
